### PR TITLE
fix(askvm): add url param to fetch

### PR DESCRIPTION
**Pull request by RichardM99**
- Resolves https://github.com/xFAANG/askql/issues/370

### CODE REVIEW

**Comment by czerwinskilukasz1** [comment]
Hello, Richard! Nice to see you here and thank you for your PR!
I am wondering if you had a chance to check that the code works as intended after your change.
The best way would be to create an .ask file with a test that calls for example `fetch('https://jsonplaceholder.typicode.com/posts/1')` (jsonplaceholder.typicode.com is a website with dummy sample data) and checks whether a value for 'id' key in the received data equals to `1`.

@RichardM99

**Comment by RichardM99** [comment]
@czerwinskilukasz1 

Thanks for the feedback - will do!

**Comment by RichardM99** [comment]
@czerwinskilukasz1 

I'm afraid I don't understand the test set up well enough to add a passing test.  I added a test (based it on the `scope-02` test) but it throws type error "fetchFun is not a function".

Also not sure whether the test should mock the http call or actually make it.

File: src/askscript/__tests__/90-misc/fetch.test.ts [hunk 1]
```diff
@@ -0,0 +1,6 @@
+import program from './fetch.ask';
+
+test('fetch', async () => {
+  const json = await program();
+  expect(json.id).toBe(1);
+});
```
Comment (mhagmajer) [inline src/askscript/__tests__/90-misc/fetch.test.ts:6]: @RichardM99 can we mock the endpoint for tests? The tests should run in isolation. You can find more about mocking in Jest here https://jestjs.io/docs/en/mock-functions

Comment (mhagmajer) [reply src/askscript/__tests__/90-misc/fetch.test.ts:6]: It's also useful to hit the real endpoint, however, that would be an integration test for Node which we don't yet have - we currently focus on unit test coverage

**Comment by czerwinskilukasz1** [comment]
> @czerwinskilukasz1
> 
> I'm afraid I don't understand the test set up well enough to add a passing test. I added a test (based it on the `scope-02` test) but it throws type error "fetchFun is not a function".

@RichardM99 , congrats, it seems you have found a bug after fixing a bug :) 
If there had been no second bug, `fetchFun` would have just worked.
How about you try to figure out why it didn't work? This Stack Overflow thread might be helpful: https://stackoverflow.com/questions/48433783/referenceerror-fetch-is-not-defined

As for mock tests which @mhagmajer suggested, let's focus on making the code compile first.

File: src/askscript/__tests__/90-misc/fetch.test.ts (hunk 1 above)
Comment (mhagmajer) [reply src/askscript/__tests__/90-misc/fetch.test.ts:6]: I'm going to let @czerwinskilukasz1 to continue guide you through your first contributions :)

**Comment by RichardM99** [comment]
@czerwinskilukasz1 

Tried both `global.fetch = require('node-fetch')` as well as just requiring it directly in `node/fetch.ts` but same error. Also not clear to me why the error is always `fetchFun` is not a function - even when I change the name of the function and `fetchFun` is not existing in the code base.

**Comment by czerwinskilukasz1** [comment]
> @czerwinskilukasz1
> 
> Tried both `global.fetch = require('node-fetch')` as well as just requiring it directly in `node/fetch.ts` but same error. Also not clear to me why the error is always `fetchFun` is not a function - even when I change the name of the function and `fetchFun` is not existing in the code base.

It's an interesting puzzle, let me take a look.

**Comment by czerwinskilukasz1** [comment]
> > @czerwinskilukasz1
> > Tried both `global.fetch = require('node-fetch')` as well as just requiring it directly in `node/fetch.ts` but same error. Also not clear to me why the error is always `fetchFun` is not a function - even when I change the name of the function and `fetchFun` is not existing in the code base.
> 
> It's an interesting puzzle, let me take a look.

OK, so I made the test work:
image

What I did is:
1. I added the following line:
image

2. Recompiled (`npm run build`) and only after recompiling did I run `npm test`.   I am guessing you just run `npm test` straightaway, which is what I sometimes do too and discover things are still failing.

OK, so now the last step.
Having:
```
global.fetch = require('node-fetch');
const fetchFun = (global as any).fetch;
```
doesn't look nice. First, we tend to use `import` construct instead of `require`. Second, I would just skip the assignment to global, so that a file deep inside the code doesn't pollute the global namespace, and would just use a local variable.

How about that, @RichardM99 ?

**Comment by RichardM99** [comment]
@czerwinskilukasz1 

That's great - thank you 🙃 

I will make the required changes over the weekend.

**Comment by RichardM99** [comment]
@czerwinskilukasz1 

Hello!

So as per your suggestions was able to make the tests pass.

However - now running into the issue of being unable to mock node-fetch.

Tried 

https://jestjs.io/docs/en/bypassing-module-mocks

And as you will see in my latest changes I also tried https://www.npmjs.com/package/jest-fetch-mock

Unfortunately the actual API call is always made instead of the mock data being resolved.

**Comment by czerwinskilukasz1** [comment]
> @czerwinskilukasz1
> 
> Hello!
> 
> So as per your suggestions was able to make the tests pass.

Good job!

> 
> However - now running into the issue of being unable to mock node-fetch.
> 
> Tried
> 
> https://jestjs.io/docs/en/bypassing-module-mocks
> 
> And as you will see in my latest changes I also tried https://www.npmjs.com/package/jest-fetch-mock
> 
> Unfortunately the actual API call is always made instead of the mock data being resolved.
> 
> EDIT: Perhaps instead of importing `node-fetch` locally in `fetch.ts` - it should be assigned globally?

Yeah, mocking can be tricky in some cases.
I like the tests you wrote though.
I couldn't make jest-fetch-mock work, it might need some extra tweaks or we are both missing something.
Anyways, I made your *first test* pass by following the example you linked: https://jestjs.io/docs/en/bypassing-module-mocks **and** in fetch.test.ts:

1. replacing
```
const result = await program();
```

with:
```
    const environment = { resources };
    const result = await e2e(
      "ask { fetch('https://jsonplaceholder.typicode.com/posts/1') }",
      environment
    );
```
and 

2. adding the following headers:
```
jest.mock('node-fetch');
import fetch from 'node-fetch';
import { resources } from '../../../askvm';
import { e2e } from '../../../utils/tools';
```

Ah, and because I used jest.mock() within fetch.test.ts, I removed setupJest.js

Why your solution didn't work?  I *suspect* importing AskQL using `import program from './fetch.ask';` uses some Jest import mechanisms and the file was read and compiled in a separate context, where fetch was not mocked.  @mhagmajer wrote the mechanism and he might be able to add more context. 
Anyways, how about you try to make my solution work for 2nd and 3rd tests too?

**Comment by RichardM99** [comment]
@czerwinskilukasz1 

Hey thanks for the comments! For sure would not have figured out to use the `e2e` function without support 🙃 

Applied your comments! Let me know what you think (maybe it's a bit weird to now have an `e2e` function in the unit test?)

**Comment by czerwinskilukasz1** [comment]
> @czerwinskilukasz1
> 
> Hey thanks for the comments! For sure would not have figured out to use the `e2e` function without support 🙃
> 

Of course, you're welcome 🙂 

> Applied your comments! Let me know what you think (maybe it's a bit weird to now have an `e2e` function in the unit test?)

As for `e2e` function, the name didn't mean to mean end-to-end testing but rather compiling and running end-to-end, meaning:
1. taking an AskScript program and environment
2. parsing it
3. transpiling to AskCode
4. executing AskCode
5. returning the result
as opposed to just parsing AskScript or just executing AskCode.
Probably we should rename it. Feel free to create an issue to this and discuss the new name there.

**Comment by RichardM99** [comment]
@czerwinskilukasz1 

I'm afraid I don't have permission to merge.

File: src/askscript/__tests__/90-misc/fetch.test.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
+jest.mock('node-fetch');
```
Comment (mhagmajer) [inline src/askscript/__tests__/90-misc/fetch.test.ts:1]: :+1:

**Comment by czerwinskilukasz1** [comment]
> @czerwinskilukasz1
    >
    > I'm afraid I don't have permission to merge.
    
    Ouch, forgot about permission. I am doing the merge now, congrats for making your first change in AskQL repository! <U+1F680>
    How did you like it? ;)

**Comment by RichardM99** [comment]
@czerwinskilukasz1 

Thanks! It was great - thanks again for the support.

Will hopefully contribute again soon.

**Comment by czerwinskilukasz1** [comment]
@RichardM99 , sounds great!  If you feel you would like any suggestion on which issue to choose, just tag me.

**Comment by RichardM99** [comment]
@czerwinskilukasz1 

If you would like to suggest an issue I will gladly work on it.

**Comment by czerwinskilukasz1** [comment]
Sounds great, @RichardM99  🙂 
How do you like this one? https://github.com/xFAANG/askql/issues/269 - it's about changing a way a resource (a Typescript function) works.
There is also https://github.com/xFAANG/askql/issues/181 , which is about adding _some_ real code examples in AskScript language.

We scheduled these issues (and a couple more) for [milestone 1.3](https://github.com/xFAANG/askql/milestone/3), which we would like to release soon.

**Comment by RichardM99** [comment]
@czerwinskilukasz1 

I can take https://github.com/xFAANG/askql/issues/269


BTW - I think when this PR was merged the commits were not squashed.

**Comment by czerwinskilukasz1** [comment]
Sounds great!

**Pull request merged**